Updating matching names (#1020)
ihmeuw/vivarium_research: docs/source/concept_models/vivarium_census_synthdata/concept_model.rst

2022_06_02b_prl_code_for_probs_of_spaces_and_hyphens_in_last_and_first_names.ipynb
for computation details.)

For now, assign all simulants within the same household who are 

> relatives of the reference person with have the same last name. 

Author: Thanks

and last name based on soundex.  Perhaps measure of success is to look
at entropy of character n-gram distribution.

> To simulate naming after a parent or family member, we would like 5% of 

Reviewer: This won't be exact, because some people drop through to the last case in the list and still are named randomly.

> To simulate naming after a parent or family member, we would like 5% of 

Author: True!

children to have the same name as a parent. This can be separated into two 
groups: 

**People born in the simulation**

For 5% of simulants, they will be assigned a name based on these steps: 

#. Female simulants will have the same last name as their mother (who is known) 
#. Male simulants, if the mother is the reference person and has an opposite-sex spouse, they will receive that name 

> #. Otherwise, if the mother has a male relative known in the simulation (either the mother is the reference person or is related to the reference person), they will be assigned a males relative's name. The relative can be selected at random. 

Author: I see your point! There definitely is some guessing at relationships here

> #. If none of these are available, assign a random name 

Author: I am sorry, I personally don't put a '.' at the end of the bullet point unless there are multiple sentences in the bullet point. This is the hill I am willing to die on although I do acknowledge that you are grammatically correct

**People initialized in the simulation**

> For 5% of simulants, they will be assigned a name based on these steps: 

Reviewer: To me, it feels fairly counter-intuitive to select the 5% of simulants *first* and then check if they can receive a matched name. But, I think it is mathematically correct.

Author: Yeah, I see what you mean. We will end up with less than 5% overall, but since the 5% was arbitrary anyways, I am not too concerned if we don't hit our made up target

> #. For anyone who has a parent of the same sex known in the simulation (this means: ["reference person", "biological child"], ["reference person", "adopted child"], ["reference person", "parent"]) they have a 5% chance of having the same name as that parent 

Author: Okay, thanks!

> #. Note that for same sex couples, whoever is the reference person will pass their name instead of their spouse 

Author: Makes sense!

> #. For anyone with a parent of the opposite sex known in the simulation (same pairings), if that parent has a spouse in the simulation of the child's sex (["reference person", "opposite sex spouse"]), they will receive the same name of the spouse 

Reviewer: ```suggestion
#. For anyone who has a child relationship attribute ("biological child", "adopted child") and is the opposite sex as the reference person, if there is someone in the household with relationship "opposite-sex spouse", they are assigned the same first name as the spouse.
```

> #. For anyone else, if their parent has a relative of the correct sex in the simulation (either the parent is the reference person or is related to the reference person), they will receive a relative's name. If there multiple relative's at random 

Reviewer: ```suggestion
#. For anyone else, if they have a relative relationship attribute (any except "roommate/housemate" and "other nonrelative"), they are assigned the same first name as another randomly-selected person in the household who also has a relative relationship attribute and the same sex. If there is no such person, skip to the next step.
```

**I think this one could give us problems.** It does not define which relative is named first. Would be curious to hear engineering's take on it. Potentially there is some way to do all of this by manipulation of seeds, which wouldn't require there to be a specific "order."

Author: Hmm, I see the issue, we are hypothetically naming sims based on sims that might not have names yet. A couple of solutions: 
1. limit the matching names to one person per household which would fix this but is likely unrealistic as I would expect families likely to name kids after a parent or relative are also likely to do this for more children 
2. Define an order - oldest to youngest might make sense? So assign 95% randomly. If a household has multiple names that are meant to be matched, select the oldest to name first, then move down through the ages. If the oldest person is meant to be named the same as a yet unnamed younger person, they just get a random name 

I am going to implement 2. Let me know if you see an issue with that.

> #. For anyone without a parent or with a parent/relatives of the wrong sex, they will be assigned a random name 

Reviewer: ```suggestion
#. For anyone else, they will be assigned a random name.
```
